### PR TITLE
support and test for generic assembler

### PR DESCRIPTION
**Pull request by n8pease**

### CODE REVIEW

File: python/lsst/daf/persistence/genericAssembler.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+import inspect
+
+def genericAssembler(dataId, componentInfo, cls):
+    """A generic assembler for butler composite datasets, that can be used when the compoent names match the
```
Comment (TallJimbo) [inline python/lsst/daf/persistence/genericAssembler.py]: typo: `compoent`

File: python/lsst/daf/persistence/genericAssembler.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    When determining setter names: If the setter name is specified by the policy then the genericAssembler
+    will use that to set the component into the python object. If the policy does not specify setter names the
+    genericAssembler will see if the __init__ func input argument names match the policy argument names. If
+    that does not work, and the python object has setter names that match the component name of all the object
+    then the setter name can be inferred; it will first try 'set' + <componentName>, and if that does not
+    exist it will try 'set' + <componentName>.capitalize (e.g. for component name 'foo', it will try setfoo
+    and then setFoo.) If no setter can be found for a component object, it will raise a runtime error.
+    """
+    args = inspect.getargspec(cls.__init__)
```
Comment (TallJimbo) [inline python/lsst/daf/persistence/genericAssembler.py]: Should guard against `TypeError` being thrown here (that's what you get if `__init__` is implemented directly in C/C++).

File: python/lsst/daf/persistence/butler.py [hunk 3]
```diff
@@ -702,11 +703,11 @@ def get(self, datasetType, dataId=None, immediate=False, **rest):
...
         if isinstance(location, ButlerComposite):
-            componentDict = {}
-            for name, info in location.componentInfo.items():
-                componentDict[name] = self.get(info.datasetType, location.dataId, immediate=True)
-            obj = location.assembler(dataId=location.dataId, componentDict=componentDict,
-                                     cls=location.python)
+            for name, componentInfo in location.componentInfo.items():
+                obj = self.get(componentInfo.datasetType, location.dataId, immediate=True)
+                componentInfo.obj = obj
+            assembler = location.assembler if location.assembler else genericAssembler
```
Comment (brianv0) [inline python/lsst/daf/persistence/butler.py]: Alternatively `assembler = location.assembler or genericAssembler`, but no big deal.

File: python/lsst/daf/persistence/butlerLocation.py [hunk 4]
```diff
@@ -74,8 +100,17 @@ def add(self, id, datasetType):
             The name of the component in the policy definition.
         datasetType : string
             The name of the datasetType of the component.
+        setter : string or None
+            The name of the function used to set this component into the python type that contains it.
+            Specifying a setter is opotional, use None if the setter won't be specified or used.
```
Comment (ktlim) [inline python/lsst/daf/persistence/butlerLocation.py]: Typo: "optional" (and below)

File: python/lsst/daf/persistence/genericAssembler.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    will use that to set the component into the python object. If the policy does not specify setter names the
+    genericAssembler will see if the __init__ func input argument names match the policy argument names. If
+    that does not work, and the python object has setter names that match the component name of all the object
+    then the setter name can be inferred; it will first try 'set' + <componentName>, and if that does not
+    exist it will try 'set' + <componentName>.capitalize (e.g. for component name 'foo', it will try setfoo
+    and then setFoo.) If no setter can be found for a component object, it will raise a runtime error.
+    """
+    canUseClassCtor = True
+    try:
+        args = inspect.getargspec(cls.__init__)
```
Comment (ktlim) [inline python/lsst/daf/persistence/genericAssembler.py]: I think you can avoid `inspect` altogether by just trying the constructor using the keyword argument dictionary and catching the exception if it fails.

File: python/lsst/daf/persistence/genericAssembler.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    useSetters = any(info.setter for info in componentInfo.values())
+    if not useSetters and canUseClassCtor and set(args.args) - set(componentInfo.keys()) == set(['self']):
+        initArgs = {k: v.obj for k, v in componentInfo.iteritems()}
+        obj = cls(**initArgs)
+    else:
+        obj = cls()
+        for componentName, componentInfo in componentInfo.iteritems():
+            if componentInfo.setter is not None:
+                setter = getattr(obj, componentInfo.setter)
+            elif hasattr(obj, 'set' + componentName):
```
Comment (ktlim) [inline python/lsst/daf/persistence/genericAssembler.py]: Given our (evolving) coding standards, it's probably better to use `'set_' + componentName`.

File: python/lsst/daf/persistence/genericAssembler.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    will use that to get the component from the python object. If the policy does not specify getter names and
+    the python object has getter names that match the component name of all the object then the getter name
+    can be inferred; it will first try 'get' + <componentName>, and if that does not exist it will try 'get' +
+    <componentName>.capitalize (e.g. for component name 'foo', it will try getfoo and then getFoo.) If no
+    getter can be found for a component object, it will raise a runtime error.
+    """
+    for componentName, componentInfo in componentInfo.iteritems():
+        if componentInfo.getter is not None:
+            getter = getattr(obj, componentInfo.getter)
+        elif hasattr(obj, 'get' + componentName):
```
Comment (ktlim) [inline python/lsst/daf/persistence/genericAssembler.py]: Same here, use `get_`.

File: python/lsst/daf/persistence/test/testObject.py [hunk 8]
```diff
@@ -44,18 +44,35 @@ class TestObjectPair(object):
     """An object for testing that contains 2 objects.
     """
 
-    def __init__(self, objA, objB):
+    def __init__(self, objA=None, objB=None):
```
Comment (ktlim) [inline python/lsst/daf/persistence/test/testObject.py]: (This really belongs in the other PR) I'm not sure you have complete test coverage of all the cases: generic assembler/disassembler with constructor, "_" setter, capitalized setter, and non-generic policy-specified assembler/disassembler

**Pull request closed without merging**